### PR TITLE
[J4] New module - Category Tags

### PR DESCRIPTION
**Pull request by korenevskiy**
### Actual result BEFORE applying this Pull Request
Previously, the display was very restrictive. Tag links opened the tag component. But the problem is that the Tag Component does not have a mode for displaying articles as a blog. At the same time, the display of the list of materials of one tag must be made additional layout.

### Expected result AFTER applying this Pull Request
The new module allows you to display tags in the form of a tree, This will allow you to use tags to display as a catalog of categories. Every developer faces a problem when an article needs to be placed in 2(3) different categories at once. For example, an article can simultaneously contain future news, as well as an article can contain information about the PHP language, as well as contain information about gifts. By opening the module tag link, a list of blog materials will be displayed.

While deleting the material tag, the material link remains the same. But if we moved an article from one category to another, the article link would change. Now the article link will remain the same, and tags can be used instead of categories.

image

UPDATED DESCRIPTION
_A known problem._
 After clicking on the tag links, the Category, Blog component stores this tag in the cache, and later, when opening any Category and Blog links, materials with the last open tag will be displayed.

_A quick solution to the problem._ 
In the link settings in the component menu, in the tag selection field, you must specify the selection of an empty tag. This will allow you to open links by resetting the tag filter in the cache.

_The required Solution to the problem._ 
It is required to make changes to the code of the Category and Blog component so that links without the selected tag reset the tag filter in the cache.

### CODE REVIEW

**Comment by brianteeman** [comment]
you dont need to include the russian language files

**Comment by brianteeman** [comment]
and is the file translate.php part of this module?

**Comment by korenevskiy** [comment]
> and is the file translate.php part of this module?

No, but this file serves as a translation for testing. I would prefer to delete it after finding the errors.

**Comment by alikon** [comment]
if you look at https://ci.joomla.org/joomla/joomla-cms/57815/1/6 there is a detailed report on what drone complain about CS
image

hope that helps

imho shouldn't be a core but a 3dp module

**Comment by korenevskiy** [comment]
@alikon  
Yes, I have been looking into this report from the very beginning. I have already removed 80% of the errors. 
But I can't fix the remaining errors. You will see for yourself the code and the error in the report. I do not understand that there is an error in an empty place. 
After that, I correct the errors as indicated in the report, and in the new report the requirement to remove the added space back. Ie, first the report asks to add a space, and in the next report asks to remove the space back. 

Maybe I don't understand something?

**Comment by brianteeman** [comment]
I would suggest not to worry for now about code style and first to get people to check what the module does and if it is something for core. _(there are lots of other style issues that drone didnt find)_

**Comment by alikon** [comment]
joomla recently switched from own cosdestyle to PSR-12 https://www.php-fig.org/psr/psr-12/
so maybe this is the main source of confusion
i've made a silly pr to your branch  to fix  only the 1st drone report 
check it maybe it will let you to reduce the cs complains

**Comment by korenevskiy** [comment]
> so maybe this is the main source of confusion

@alikon 
You're right, the problem was in the curly brace after the condition. But in the error report, it is NOT stated that the problem is curly brace. The report just says about the brace. And since the curly brace is on another line, that implies a parenthesis, as well as the error is indicated in the absence of a space. The error in the report does not indicate the cause of the error at all.

**Thank you for your help, now I will know about the new code format.**

**Comment by korenevskiy** [comment]
`modules/mod_category_tags/tmpl/_childs.php (line 1)`
@alikon 
I have fixed all the errors, there is one more error that I cannot remove. Please help me fix it. The parser swears that the line break "\r\n" is used instead of "\n". No matter how much I would try to fix the transfer, the error remains in the code.

**Comment by richard67** [comment]
@korenevskiy Regarding naming of files and variables: The plural of "child" in English language is "children", not "childs".

**Comment by richard67** [comment]
> `modules/mod_category_tags/tmpl/_childs.php (line 1)` @alikon I have fixed all the errors, there is one more error that I cannot remove. Please help me fix it. The parser swears that the line break "\r\n" is used instead of "\n". No matter how much I would try to fix the transfer, the error remains in the code.

@korenevskiy You could use a good editor like notepad++ which can show you what kind of line feeds are in the file (Unix like it should be, or Windows like the PHP CS check complains about).

File: modules/mod_category_tags/mod_category_tags.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_category_tags
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (Kostelano) [inline modules/mod_category_tags/mod_category_tags.php]:
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```

File: modules/mod_category_tags/mod_category_tags.xml [hunk 2]
```diff
@@ -0,0 +1,252 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension type="module" client="site" method="upgrade">
+	<name>mod_category_tags</name>
+	<author>Joomla! Project</author>
+	<creationDate>2022-05-09</creationDate>
+	<copyright>(C) 2013 Open Source Matters, Inc.</copyright>
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>4.0.0</version>
```
Comment (Kostelano) [inline modules/mod_category_tags/mod_category_tags.xml]:
```suggestion
	<creationDate>2022-05-09</creationDate>
	<copyright>(C) 2022 Open Source Matters, Inc.</copyright>
	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
	<authorEmail>admin@joomla.org</authorEmail>
	<authorUrl>www.joomla.org</authorUrl>
	<version>__DEPLOY_VERSION__</version>
```

File: language/en-GB/mod_category_tags.ini [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+MOD_CATEGORY_TAGS_FIELD_MINSIZE_LABEL="Minimum Font Size"
+MOD_CATEGORY_TAGS_FIELD_NO_RESULTS_LABEL="Show &quot;No results&quot; text"
+MOD_CATEGORY_TAGS_FIELD_NO_RESULTS_TEXT = "No results message"
+MOD_CATEGORY_TAGS_NO_ITEMS_FOUND="No Tags found."
+MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_COUNT="Number of Items"
+MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_LABEL="Order"
+MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_RANDOM="Random"
+MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_TITLE="Title"
+MOD_CATEGORY_TAGS_FIELD_TIMEFRAME_LABEL="Time Period"
+MOD_CATEGORY_TAGS_FIELD_DISPLAY_TREE_LABEL ="Show tree"
```
Comment (Kostelano) [inline language/en-GB/mod_category_tags.ini]:
```suggestion
; (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
; License GNU General Public License version 2 or later; see LICENSE.txt
; Note : All ini files need to be saved as UTF-8

MOD_CATEGORY_TAGS="Category - Tags"
MOD_CATEGORY_TAGS_FIELD_ALL_TIME="All time"
MOD_CATEGORY_TAGS_FIELD_DISPLAY_COUNT_LABEL="Display Number of Items"
MOD_CATEGORY_TAGS_FIELD_LAST_DAY="Last day"
MOD_CATEGORY_TAGS_FIELD_LAST_HOUR="Last hour"
MOD_CATEGORY_TAGS_FIELD_LAST_MONTH="Last month"
MOD_CATEGORY_TAGS_FIELD_LAST_WEEK="Last week"
MOD_CATEGORY_TAGS_FIELD_LAST_YEAR="Last year"
MOD_CATEGORY_TAGS_FIELD_MAXSIZE_DESC="The maximum font size used for the tags, proportional to the site's default font size (eg \"2\" means 200% of the default size)."
MOD_CATEGORY_TAGS_FIELD_MAXSIZE_LABEL="Maximum Font Size"
MOD_CATEGORY_TAGS_FIELD_MINSIZE_DESC="The minimum font size used for the tags, proportional to the site's default font size (eg \"2\" means 200% of the default size)."
MOD_CATEGORY_TAGS_FIELD_MINSIZE_LABEL="Minimum Font Size"
MOD_CATEGORY_TAGS_FIELD_NO_RESULTS_LABEL="Show \"No results\" text"
MOD_CATEGORY_TAGS_FIELD_NO_RESULTS_TEXT="No results message"
MOD_CATEGORY_TAGS_NO_ITEMS_FOUND="No Tags found."
MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_COUNT="Number of Items"
MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_LABEL="Order"
MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_RANDOM="Random"
MOD_CATEGORY_TAGS_FIELD_ORDER_VALUE_TITLE="Title"
MOD_CATEGORY_TAGS_FIELD_TIMEFRAME_LABEL="Time Period"
MOD_CATEGORY_TAGS_FIELD_DISPLAY_TREE_LABEL="Show tree"
```

File: language/en-GB/mod_category_tags.sys.ini [hunk 4]
```diff
@@ -0,0 +1,9 @@
+; Joomla! Project
+; (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
+; License GNU General Public License version 2 or later; see LICENSE.txt
+; Note : All ini files need to be saved as UTF-8
+
+MOD_CATEGORY_TAGS="Category - tags"
```
Comment (Kostelano) [inline language/en-GB/mod_category_tags.sys.ini]:
```suggestion
; (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
; License GNU General Public License version 2 or later; see LICENSE.txt
; Note : All ini files need to be saved as UTF-8

MOD_CATEGORY_TAGS="Category - Tags"
```

File: modules/mod_category_tags/src/Helper/CategoryTagsHelper.php [hunk 5]
```diff
@@ -0,0 +1,198 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_category_tags
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (Kostelano) [inline modules/mod_category_tags/src/Helper/CategoryTagsHelper.php]:
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```

File: modules/mod_category_tags/tmpl/_childs.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_category_tags
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (Kostelano) [inline modules/mod_category_tags/tmpl/_childs.php]:
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>

```

File: modules/mod_category_tags/tmpl/cloud.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_category_tags
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (Kostelano) [inline modules/mod_category_tags/tmpl/cloud.php]:
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```

File: modules/mod_category_tags/tmpl/default.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_category_tags
+ *
+ * @copyright   (C) 2013 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (Kostelano) [inline modules/mod_category_tags/tmpl/default.php]:
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```

**Review by Kostelano — CHANGES_REQUESTED** [review]

File: modules/mod_category_tags/language/en-GB/mod_category_tags.ini [hunk 9]
```diff
@@ -0,0 +1,40 @@
+; Joomla! Project
```
Comment (Kostelano) [inline modules/mod_category_tags/language/en-GB/mod_category_tags.ini]: This file and the "language" directory must be deleted, no localization files are stored in the modules directory.

File: modules/mod_category_tags/mod_category_tags.xml (hunk 2 above)
Comment (korenevskiy) [reply modules/mod_category_tags/mod_category_tags.xml]: Why do we need to write the "creationDate" not in SQL format?

Comment (Kostelano) [reply modules/mod_category_tags/mod_category_tags.xml]: [Joomla recently switched to this date format](https://github.com/joomla/joomla-cms/pull/27162). You can check in any module.

File: modules/mod_category_tags/mod_category_tags.xml [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+					mode="nested"
+				/>
+
+				<field
+					name="category_layout"
+					type="list"
+					label="MOD_CATEGORY_TAGS_FIELD_CATEGORY_LAYOUT_LABEL"
+					default="blog"
+					validate="options"
+				>
```
Comment (brianteeman) [inline modules/mod_category_tags/mod_category_tags.xml:50]: Please fix the indentation here and in all similar areas.

File: modules/mod_category_tags/mod_category_tags.xml [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+		<folder>language</folder>
+		<folder>src</folder>
+		<folder>tmpl</folder>
+	</files>
+	<config>
+		<fields name="params">
+			<fieldset name="basic">
+				
+				<field
+					type="translate"/>
```
Comment (brianteeman) [inline modules/mod_category_tags/mod_category_tags.xml]: ???

File: modules/mod_category_tags/mod_category_tags.xml [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+					<option value="0">JGLOBAL_ORDER_ASCENDING</option>
+					<option value="1">JGLOBAL_ORDER_DESCENDING</option>
+				</field>
+
+
+				<field
+					name="tree_display"
+					type="radio"
+					layout="joomla.form.field.radio.switcher"
+					label="MOD_CATEGORY_TAGS_FIELD_DISPLAY_TREE_LABEL"
```
Comment (brianteeman) [inline modules/mod_category_tags/mod_category_tags.xml:108]: name
type
label
layout

please fix this ordering here and in all similar areas

File: modules/mod_category_tags/mod_category_tags.xml [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+					<option value="1">JYES</option>
+				</field>
+
+				<field
+					name="title_display"
+					type="radio"
+					layout="joomla.form.field.radio.switcher"
+					label="MOD_CATEGORY_TAGS_FIELD_DISPLAY_TITLE_LABEL"
+					default="1"
+					filter="integer"
```
Comment (brianteeman) [inline modules/mod_category_tags/mod_category_tags.xml:122]: filter="boolean"

**Comment by Kostelano** [comment]
By the way, replace the branch with 4.3.

File: modules/mod_category_tags/mod_category_tags.xml (hunk 2 above)
Comment (korenevskiy) [reply modules/mod_category_tags/mod_category_tags.xml]: > [Joomla recently switched to this date format](https://github.com/joomla/joomla-cms/pull/27162). You can check in any module.

This is a big mistake of all modules. An XML file is a data file just like SQL is designed for data. The XML file is not a text document. We store data in SQL in ISO format. Maybe we will store data in SQL in the "Nov 2022" format?

Thank you so much. I have made all your corrections except for the CreateionDate tag.

Comment (Kostelano) [reply modules/mod_category_tags/mod_category_tags.xml]: I proposed a correction based on the previously adopted PR (for compliance). But most likely you are right and this is a mistake. There is already at least an https://github.com/joomla/joomla-cms/issues/38552 on this topic.

Comment (Kostelano) [reply modules/mod_category_tags/mod_category_tags.xml]: I have edited the change request. The date will need to be discussed with the release leaders.

**Comment by korenevskiy** [comment]
Dear colleagues. I have created a new PR for branch 4.3. I am not able to transfer PR to a new branch. I decided to create anew. Here's a new PR https://github.com/joomla/joomla-cms/pull/38752

@brianteeman , @Kostelano  , @chmst  , @alikon  ,  @richard67  , @laoneo  , @rdeutz  , @Hackwar  , @wilsonge  , @zero-24  , @bembelimen , @joomdonation  , @obuisard  , @HLeithner  , @heelc29  , @sonvnn @webuniverse

File: modules/mod_category_tags/mod_category_tags.xml (hunk 13 above)
Comment (korenevskiy) [reply modules/mod_category_tags/mod_category_tags.xml:122]: in new PR changed

File: modules/mod_category_tags/mod_category_tags.xml (hunk 12 above)
Comment (korenevskiy) [reply modules/mod_category_tags/mod_category_tags.xml:108]: in new PR changed

File: modules/mod_category_tags/mod_category_tags.xml (hunk 2 above)
Comment (korenevskiy) [reply modules/mod_category_tags/mod_category_tags.xml]: > There is already at least an #38552 on this topic.

Thank you, I indicated there in the message that XML is a data format.

**Pull request closed without merging**